Make cleanup phase 1

ProjectOblivion opened the pull request:
In an effort to clean up the build pipeline:
I moved most of the build targets to more appropriate locations.  Common targets in Makefile.tools.mk and platform specific targets in their respective platform .mk files.
I moved the variables and functions to Makefile for easier editing.  All of the build targets are now at the top of the Makefile.
I moved the function merge lists out of the psp makefile and into files in the config/ dir.  The variable assignments are in the main Makefile in a psp specific block.

The goal is to have a common predictable pattern for all platforms, to minimize the need to edit the makefiles, and to make editing straightforward if editing the files is needed.  This isn't even close to what I would consider an end, but it is a start.

sozud commented on Makefile:
```diff
@@ -77,7 +107,86 @@ PNG2S           := $(PYTHON) $(TOOLS_DIR)/png2s.py
+ST_WRP_MERGE := $(file < $(WRP_MERGE_FILE))
+ST_NO4_MERGE := $(file < $(NO4_MERGE_FILE))
+endif
+#else ifeq ($(VERSION),saturn) Workaround for github actions
```
is this supposed to be here? same for the #endif later

sozud reviewed (approved):
Seems fine but @Xeeynamo ought to approve too

hohle commented:
Some additional questions:

Does this fix any relationships between symbols, splat config, and built overlays? one of the pain points for me is missing dependencies between the recipes which forces frequent cleans between rebases, segment splits, etc.

Does this impact build times at all with parallel builds? I didn't see anything that looked like it would, but wanted to double check.

ProjectOblivion commented on Makefile:
```diff
@@ -77,7 +107,86 @@ PNG2S           := $(PYTHON) $(TOOLS_DIR)/png2s.py
+MWCCGAP_APP     := $(MWCCGAP_DIR)/mwccgap.py
+MWCCGAP         := $(PYTHON) $(MWCCGAP_APP)
+
+DRA_MERGE_FILE := $(CONFIG_DIR)/merge.dra.txt
+LIB_MERGE_FILE := $(CONFIG_DIR)/merge.lib.txt
+ST0_MERGE_FILE := $(CONFIG_DIR)/merge.st0.txt
+WRP_MERGE_FILE := $(CONFIG_DIR)/merge.wrp.txt
+NO4_MERGE_FILE := $(CONFIG_DIR)/merge.no4.txt
+
+ST_DRA_MERGE := $(file < $(DRA_MERGE_FILE)) 
+ST_LIB_MERGE := $(file < $(LIB_MERGE_FILE))
+ST_ST0_MERGE := $(file < $(ST0_MERGE_FILE))
+ST_WRP_MERGE := $(file < $(WRP_MERGE_FILE))
+ST_NO4_MERGE := $(file < $(NO4_MERGE_FILE))
```
It was based primarily on the expectation that some of these lists will grow quite cumbersome over time, but also because they are not static until the overlay is complete.  This allows those functions to be added without the makefiles being edited.

ProjectOblivion replied:
I will revert the file changes back to inline for now and I'll look for alternative ways to handle the merge step.

ProjectOblivion replied:
Sorry, I can see how that wasn't clear.  I meant exactly what you said, reverting the changes now and looking for a better way for a future PR.

ProjectOblivion commented on Makefile.tools.mk:
```diff
@@ -0,0 +1,418 @@
+.PHONY: build
+
+##@ Variables
+##@
+##@     VERSION              the game version to build (us, hd, pspeu, saturn, pc) (Default: us)
+##@
+##@ Primary Targets
+##@
+
+all: ##@ (Default) build and check
+all: build check
+
+extract: ##@ split game files into assets and assembly
+extract: extract_$(VERSION)
+
+build: ##@ build game files
+build: build_$(VERSION)
```
While I can appreciate your point of view, it seems unnecessary considering there is a help target for make, as well as the documentation in the wiki.  Even before I touched the makefile, there really wasn't a need to look for the targets in it.

ProjectOblivion replied:
For clarification, what are we calling "default targets"?  My interpretation would be `build` and `check`, the two prerequisites defined by `all`, as well as `all` itself.

ProjectOblivion commented:
> Does this fix any relationships between symbols, splat config, and built overlays? one of the pain points for me is missing dependencies between the recipes which forces frequent cleans between rebases, segment splits, etc.

It doesn't yet, but that is one of the end goals of this project.  I'm aiming to have it only rebuild what was actually touched, instead of a clean build every time.  I have issue #2195 if you want to add those or any additional things to it.

> Does this impact build times at all with parallel builds? I didn't see anything that looked like it would, but wanted to double check.

The current changes do not, a primary goal of mine to at a minimum not increase build times.  In my mind, performance takes precedence over pretty.

ProjectOblivion commented on Makefile:
```diff
@@ -77,7 +107,86 @@ PNG2S           := $(PYTHON) $(TOOLS_DIR)/png2s.py
+CHECK_FILES := $(shell cut -d' ' -f3 config/check.$(VERSION).sha)
+
+# Platform specific variables
+ifneq (,$(filter $(VERSION),us hd)) # Both us and hd versions use the PSX platform
```
Yeah, I have no arguments against that.  At the end of the day, the goal is to make it easier to work with overall and not just easier for me to work with it.  It was a bit difficult since things are so tangled, like you said, so it will become easier over time to clean up many of these things.

ProjectOblivion commented on Makefile.psp.mk:
```diff
@@ -68,47 +15,42 @@ $(MWCCGAP_APP):
 	$(OBJCOPY) -O binary $< $@
 $(BUILD_DIR)/wrp.bin: $(BUILD_DIR)/stwrp.elf
 	$(OBJCOPY) -O binary $< $@
+$(BUILD_DIR)/no4.bin: $(BUILD_DIR)/stno4.elf
+	$(OBJCOPY) -O binary $< $@
```
Agreed, that was my oversight when I added it locally.

ProjectOblivion commented on Makefile.psx.mk:
```diff
@@ -117,8 +102,167 @@ $(BUILD_DIR)/assets/st/sel/memcard_%.png.o: assets/st/sel/memcard_%.png
+$(BUILD_DIR)/F_RWRP.BIN:
+	$(GFXSTAGE) e assets/st/rwrp $@
+
+no4: $(BUILD_DIR)/NO4.BIN $(BUILD_DIR)/F_NO4.BIN
+$(BUILD_DIR)/NO4.BIN: $(BUILD_DIR)/stno4.elf
+	$(OBJCOPY) -O binary $< $@
+$(BUILD_DIR)/F_NO4.BIN:
+	$(GFXSTAGE) e assets/st/no4 $@
```
Same as the other one, my oversight.

ProjectOblivion commented on Makefile.tools.mk:
```diff
@@ -0,0 +1,418 @@
+##@ Primary Targets
+##@
+
+all: ##@ (Default) build and check
+all: build check
```
While this is the first goal to appear as far as make is concerned and also not the only method for ensuring a default goal, I can see the potential issues with it being an included file.
However, it is also not the only way to ensure the default goal.  Setting the .DEFAULT_GOAL variable also ensures that make with no arguments will call it.

ProjectOblivion commented:
I reverted the psp merge files, addressed the no4 positioning, and removed the workarounds.  I think I got all of the comments that needed to be addressed for merging, but please let me know if I missed anything.

To get many of the build targets back into Makefile I simply copy and pasted the tools file into the import location.  I did this to in the hopes that it would make a second review easier, but it ended up with some slightly odd ordering.  Functionally it should not be an issue, but I can adjust things if needed.
